File: House_Robber_III.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var rob = function(root) {
    if (!root) {
        return 0;
    }
    
    function dfs(node) {
        if (!node) {
            return [0, 0];
        }
        
        const left = dfs(node.left);
        const right = dfs(node.right);
        
        const robCur = node.val + left[1] + right[1];
        const notRobCur = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
        
        return [robCur, notRobCur];
    }
    
    const res = dfs(root);
    return Math.max(res[0], res[1]);    
};

File: House_Robber_III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        def dfs(node):
            if not node:
                return [0, 0]
            
            left = dfs(node.left)
            right = dfs(node.right)

            rob_cur = node.val + left[1] + right[1]
            not_rob_cur = max(left) + max(right)

            return [rob_cur, not_rob_cur]

        res = dfs(root)
        return max(res)        

# Time : O(n)
# Space : O(h)        
